File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/laser_scan.py
"""Generated mapper for `sensor_msgs/msg/LaserScan`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.sensor_msgs.msg.v1 import LaserScan as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def laser_scan_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.angle_min = msg.angle_min
    bus.angle_max = msg.angle_max
    bus.angle_increment = msg.angle_increment
    bus.time_increment = msg.time_increment
    bus.scan_time = msg.scan_time
    bus.range_min = msg.range_min
    bus.range_max = msg.range_max
    bus.ranges.extend(list(msg.ranges))
    bus.intensities.extend(list(msg.intensities))
    return bus


def laser_scan_to_ros(bus):
    from sensor_msgs.msg import LaserScan as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.angle_min = bus.angle_min
    out.angle_max = bus.angle_max
    out.angle_increment = bus.angle_increment
    out.time_increment = bus.time_increment
    out.scan_time = bus.scan_time
    out.range_min = bus.range_min
    out.range_max = bus.range_max
    out.ranges = list(bus.ranges)
    out.intensities = list(bus.intensities)
    return out


class SensorMsgsLaserScanMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from sensor_msgs.msg import LaserScan as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return laser_scan_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return laser_scan_to_ros(bus)

File: src/ros2_bridge/mappers/sensor_msgs/laser_scan.rs
//! Typed mapper for `sensor_msgs/msg/LaserScan`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn laser_scan_to_bus(
    msg: ros_env::sensor_msgs::msg::LaserScan,
) -> crate::sensor_msgs::msg::v1::LaserScan {
    crate::sensor_msgs::msg::v1::LaserScan {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        angle_min: msg.angle_min,
        angle_max: msg.angle_max,
        angle_increment: msg.angle_increment,
        time_increment: msg.time_increment,
        scan_time: msg.scan_time,
        range_min: msg.range_min,
        range_max: msg.range_max,
        ranges: crate::ros2_bridge::mappers::convert::f32_seq(msg.ranges),
        intensities: crate::ros2_bridge::mappers::convert::f32_seq(msg.intensities),
    }
}

pub(crate) fn laser_scan_to_ros(
    bus: crate::sensor_msgs::msg::v1::LaserScan,
) -> ros_env::sensor_msgs::msg::LaserScan {
    ros_env::sensor_msgs::msg::LaserScan {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(
            bus.header.unwrap_or_default(),
        ),
        angle_min: bus.angle_min,
        angle_max: bus.angle_max,
        angle_increment: bus.angle_increment,
        time_increment: bus.time_increment,
        scan_time: bus.scan_time,
        range_min: bus.range_min,
        range_max: bus.range_max,
        ranges: bus.ranges,
        intensities: bus.intensities,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsLaserScanMapper;

impl TypedTopicMapper for SensorMsgsLaserScanMapper {
    type Ros = ros_env::sensor_msgs::msg::LaserScan;
    type Bus = crate::sensor_msgs::msg::v1::LaserScan;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(laser_scan_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(laser_scan_to_ros(msg))
    }
}
